engine/access/index/engine.go

```diff
@@ -0,0 +1,293 @@
+	// fail early if no notification has been received for the given block height.
+	// note: it's possible for the data to exist in the data store before the notification is
+	// received. this ensures a consistent view is available to all callers.
+	if height > e.highestHeight.Value() {
+		return nil, fmt.Errorf("execution data for block %d is not available yet: %w", height, storage.ErrNotFound)
+	}
+
+	execData, err := e.execDataCache.ByHeight(ctx, height)
+	if err != nil {
+		return nil, fmt.Errorf("could not get execution data for block %d: %w", height, err)
+	}
+
+	return execData, nil
+}
+
+func (e *Engine) handleExecutionData(execData *execution_data.BlockExecutionDataEntity) error {
+	// TODO: should we parallelize indexing collections and events for each chunk?
+	for i, chunkData := range execData.ChunkExecutionDatas {
+		err := e.handleCollection(execData.BlockID, chunkData.Collection)
+		if err != nil {
+			return fmt.Errorf("could not index collection for chunk %d: %w", i, err)
+		}
+
+		err = e.handleEvents(execData.BlockID, chunkData.Events)
+		if err != nil {
+			return fmt.Errorf("could not index events for chunk %d: %w", i, err)
+		}
+
+		err = e.handleTrieUpdate(execData.BlockID, chunkData.TrieUpdate)
+		if err != nil {
+			return fmt.Errorf("could not index trie update for chunk %d: %w", i, err)
+		}
+	}
+
+	return nil
+}
+
+func (e *Engine) handleCollection(blockID flow.Identifier, collection *flow.Collection) error {
+	collID := collection.ID()
+	light := collection.Light()
+
+	lg := e.log.With().
+		Hex("block_id", logging.ID(blockID)).
+		Hex("collection_id", logging.ID(collID)).
+		Logger()
+
+	// if we haven't already indexed the collection, index it now
+	// this is an optimization while we're still using the ingestion engine for collections.
+	found, err := e.lookupCollection(collID)
+	if err != nil {
+		return err
+	}
+
+	// collection already indexed, nothing to do
+	if found {
+		return nil
+	}
+
+	// store the light collection (collection minus the transaction body - those are stored separately)
+	// and add transaction ids as index
+	err = e.collections.StoreLightAndIndexByTransaction(&light)
+	if err != nil {
+		// ignore collection if already seen
+		if errors.Is(err, storage.ErrAlreadyExists) {
+			lg.Debug().Msg("collection is already seen")
+			return nil
+		}
+		return err
+	}
+
+	// now store each of the transaction body
+	for _, tx := range collection.Transactions {
+		err := e.transactions.Store(tx)
+		if err != nil {
+			return fmt.Errorf("could not store transaction (%x): %w", tx.ID(), err)
+		}
+	}
+
+	return nil
+}
+
+func (e *Engine) handleEvents(blockID flow.Identifier, events flow.EventsList) error {
+	// Note: service events are currently not included in execution data
+	// see https://github.com/onflow/flow-go/issues/4624
+	return e.events.Store(blockID, []flow.EventsList{events})
```

Comment:
Note: the last chunk in an execution data is the system chunk. All events in that `ChunkExecutionData` are service events.